### DISCUSSION TITLE
My project binary could not find libmpi.so.12

### DISCUSSION THREAD

**Discussion by giuseppe998e** [rsmpi/rsmpi#156] [Q&A] [answered]
Hi, I am currently trying to run the _rsmpi_ `time.rs` example on a machine with Spark installed. To do this, I have installed the following packages: `rust@1.65.0`, `llvm@12.0.1`, `openmpi@4.1.5` and `mpich@4.0.2`. However, when I try to run cargo run for the project, I encounter the following error:
```
target/release/rust_openmpi_test: error while loading shared libraries: libmpi.so.12: cannot open shared object file: No such file or directory
```

I checked my system and found that I only have `libmpi.so.40`.  
And from what I can gather, the library mentioned in the error is at least six years old.
  
Do you know how I can solve it?

**Comment by jedbrown** [comment, marked as answer]
If you have mpich installed, then there is also `libmpi.so.12`. The problem is that detection found the build toolchain for MPICH, but the MPICH directories are not in your default search path. Probably setting `LD_LIBRARY_PATH` to point at your `/path/to/mpich/lib` will allow you to run. Alternatively, you can `cargo clean` and make sure it's the Open MPI tools that get found.

**Reply by giuseppe998e** [reply]
I can confirm that Spack does not add the `libmpi.so.12' library from `mpich@4.0.2' to the `LD_LIBRARY_PATH' as it should.  
If I find how to fix it I will update the answer, maybe it could be helpful to others, thanks!

**Reply by jedbrown** [reply]
You can use Spack [environments](https://spack.readthedocs.io/en/latest/environments.html) or modules for such things. It depends how much spack you want in your food.

**Reply by giuseppe998e** [reply]
Oh thank you for the clarification, I am learning it, I have to stay with Spack.
